### PR TITLE
Add support to 'getagentsstats' command in remoted and analysisd

### PR DESCRIPTION
**Pull request by sdvendramini**
|Related issue|
|---|
https://github.com/wazuh/wazuh/issues/14209
||

## Description

This PR add changes to implement a command in `wazuh-remoted` and `wazuh-analysisd` to return the stats for agents.

The name of the command will be `getagentsstats`. It will also accept two parameters:

- `agents`: An array with the ID of the agents to query stats.
  - This parameter will also accept the keyword `all` to query all agents.
- `last_id`: To allow pagination when using the keyword `all` since the response could be very large (over 64KB), this parameter is required to indicate the ID of the last agent retrieved in the previous query.
  - The first query using `all` has to set this parameter to 0.
  - When there are more agents to query, the response will have a particular error message (`due`) to indicate this.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade

- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [x] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/wazuh_db/wdb_global.c [hunk 1]
```diff
@@ -1822,6 +1822,11 @@ cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id
             return NULL;
         }
     }
+    if (sqlite3_bind_int(stmt, 4, limit) != SQLITE_OK) {
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/wdb_global.c:1825]: Done! :heavy_check_mark:

File: src/wazuh_db/wdb.h [hunk 2]
```diff
@@ -2097,7 +2097,7 @@ int wdb_global_reset_agents_connection(wdb_t *wdb, const char *sync_status);
  * @retval JSON with agents IDs on success.
  * @retval NULL on error.
  */
-cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status, const char* node_name);
+cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status, const char* node_name, int limit);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/wdb.h:2100]: Done!

File: src/wazuh_db/helpers/wdb_global_helpers.h [hunk 3]
```diff
@@ -280,4 +280,15 @@ int* wdb_disconnect_agents(int keepalive, const char *sync_status, int *sock);
...
+/**
+ * @brief Get agents by status.
+ *
+ * @param[in] connection_status Filter the query by agent connection status.
+ * @param[in] last_id Filter the query whit ids higer than this value.
+ * @param[out] count Store number of rows returned.
+ * @param[in] limit Limit number of rows returned.
+ * @return Returns an array of agents ids.
+ */
+int* get_connected_agents_ids(const char* connection_status, int last_id, int *count, int limit);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.h:292]: Name Changed!

File: src/remoted/remcom.c [hunk 4]
```diff
@@ -22,21 +22,29 @@
...
 
 const char * error_messages[] = {
     [ERROR_OK] = "ok",
+    [ERROR_DUE] = "due",
     [ERROR_INVALID_INPUT] = "Invalid JSON input",
     [ERROR_EMPTY_COMMAND] = "Empty command",
     [ERROR_UNRECOGNIZED_COMMAND] = "Unrecognized command",
     [ERROR_EMPTY_PARAMATERS] = "Empty parameters",
     [ERROR_EMPTY_SECTION] = "Empty section",
+    [ERROR_EMPTY_AGENTS] = "Empty agents", // esste no seria mejor algo como: invalid agents try [ 1 , 2 , ... , N] or "all" ?????
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/remcom.c:45]: Done!

File: src/remoted/remcom.c [hunk 5]
```diff
@@ -69,7 +77,7 @@ STATIC char* remcom_output_builder(int error_code, const char* message, cJSON* d
     cJSON* root = cJSON_CreateObject();
 
     cJSON_AddNumberToObject(root, "error", error_code);
-    cJSON_AddStringToObject(root, "message", message);
+    cJSON_AddStringToObject(root, "message", message);  // este no pasa al final del json?
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/remcom.c:80]: :+1:

File: src/remoted/state.h [hunk 6]
```diff
@@ -196,8 +199,16 @@ cJSON* rem_create_state_json();
...
  */
-cJSON* rem_create_agents_state_json();
+cJSON* rem_create_agents_state_json(int* agents_ids);
+
+/**
+ * @brief Create a int array from CJson array
+ * @param array_json CJson array object to convert.
+ * @return Int array
+ */
+int* cjson_to_array(cJSON * array_json);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/state.h:212]: Done!

File: src/remoted/remcom.c [hunk 7]
```diff
@@ -95,7 +107,43 @@ STATIC size_t remcom_dispatch(char* request, char** output) {
...
+            if (parameters_json = cJSON_GetObjectItem(request_json, "parameters"), cJSON_IsObject(parameters_json)) {
+                agents_json = cJSON_GetObjectItem(parameters_json, "agents");
+                if (cJSON_IsArray(agents_json)) {
+                    if (cJSON_GetArraySize(agents_json) <  MAX_NUM_AGENTS_STATS) {
+                        agents_ids = cjson_to_array(agents_json);
+                        if (agents_ids != NULL) {
+                            *output = remcom_output_builder(ERROR_OK, error_messages[ERROR_OK], rem_create_agents_state_json(agents_ids));
+                            os_free(agents_ids);
+                        } else {
+                            *output = remcom_output_builder(ERROR_EMPTY_AGENTS, error_messages[ERROR_EMPTY_AGENTS], NULL);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/remcom.c:119]: For both cases 'all' and with array of ids the response is:
```json
{'error': 0, 'message': 'ok', 'data': {'timestamp': 1658250869, 'name': 'wazuh-remoted', 'agents': []}}
```

File: src/analysisd/state.c [hunk 8]
```diff
@@ -529,9 +529,8 @@ STATIC analysisd_agent_state_t * get_node(const char *agent_id) {
 }
 
 STATIC void w_analysisd_clean_agents_state() {
-    char *node_name = NULL;
     int *active_agents = NULL;
-    int sock = -1;
+    int count;
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/analysisd/state.c:533]: Done!

File: src/remoted/state.c [hunk 9]
```diff
@@ -228,9 +227,8 @@ STATIC remoted_agent_state_t * get_node(const char *agent_id) {
 }
 
 STATIC void w_remoted_clean_agents_state() {
-    char *node_name = NULL;
     int *active_agents = NULL;
-    int sock = -1;
+    int count;
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/state.c:231]: Done!

File: src/remoted/state.c [hunk 10]
```diff
@@ -637,37 +625,56 @@ cJSON* rem_create_agents_state_json() {
...
+    int i;
+
+    if (array_size == 0) {
+        return NULL;
+    }
+
+    os_calloc(array_size + 1, sizeof(int), array);
+
+    for (i = 0; i < array_size; i++) {
+        array[i] = cJSON_GetArrayItem(array_json, i)->valueint;
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/state.c:672]: Function implementation changed.

File: src/remoted/state.c [hunk 11]
```diff
@@ -603,28 +597,22 @@ cJSON* rem_create_state_json() {
...
 
-    if (hash_node = OSHash_Begin(remoted_agents_state, &index), hash_node != NULL) {
-        remoted_agent_state_t * data = NULL;
-
-        cJSON *_array = cJSON_CreateArray();
-
-        while (hash_node != NULL) {
-            data = hash_node->data;
-
+    for (int i = 0; agents_ids[i] != -1; i++) {
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/remoted/state.c:611]: Done!

File: src/shared/json_op.c [hunk 12]
```diff
@@ -114,3 +114,39 @@ void json_strip(char * json) {
         }
     }
 }
+
+int* json_parse_agents(const cJSON* agents) {
+    char *output = NULL;
```
Comment (TomasTurina) [inline src/shared/json_op.c:119]: `output` is not necessary

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/wazuh_db/helpers/wdb_global_helpers.h [hunk 13]
```diff
@@ -280,4 +279,16 @@ int* wdb_disconnect_agents(int keepalive, const char *sync_status, int *sock);
...
+ * @brief Returns an array containing the ID of every agent (excluding the manager) that matches
+ *        the specified connection status, last_id and limit, ended with -1.
+ *
+ * @param[in] connection_status Filter the query by agent connection status.
+ * @param[in] last_id Filter the query whit ids higer than this value.
+ * @param[out] count Store number of rows returned.
+ * @param[in] limit Limit number of rows returned.
+ * @return Returns pointer to the array of agents ids, on success. NULL on errors.
+ */
+int* wdb_get_agents_ids_of_current_node(const char* connection_status, int *sock, int last_id, int *count, int limit);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.h:292]: Done!

File: src/wazuh_db/helpers/wdb_global_helpers.h [hunk 14]
```diff
@@ -280,4 +279,16 @@ int* wdb_disconnect_agents(int keepalive, const char *sync_status, int *sock);
  */
 time_t get_agent_date_added(int agent_id);
 
+/**
+ * @brief Returns an array containing the ID of every agent (excluding the manager) that matches
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.h:283]: Done!

File: src/shared/json_op.c [hunk 15]
```diff
@@ -114,3 +114,39 @@ void json_strip(char * json) {
...
+
+    if (error_flag) {
+        os_free(agent_ids);
+        return NULL;
+    }
+
+    os_free(output);
+
+    return agent_ids;
+}
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/shared/json_op.c:152]: Done!

File: src/wazuh_db/wdb.c [hunk 16]
```diff
@@ -174,8 +174,8 @@ static const char *SQL_STMT[] = {
...
     [WDB_STMT_GLOBAL_UPDATE_AGENT_INFO] = "UPDATE agent SET config_sum = :config_sum, ip = :ip, manager_host = :manager_host, merged_sum = :merged_sum, name = :name, node_name = :node_name, os_arch = :os_arch, os_build = :os_build, os_codename = :os_codename, os_major = :os_major, os_minor = :os_minor, os_name = :os_name, os_platform = :os_platform, os_uname = :os_uname, os_version = :os_version, version = :version, last_keepalive = :last_keepalive, connection_status = :connection_status, disconnection_time = :disconnection_time, group_config_status = :group_config_status, sync_status = :sync_status WHERE id = :id;",
     [WDB_STMT_GLOBAL_GET_AGENTS] = "SELECT id FROM agent WHERE id > ?;",
-    [WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS] = "SELECT id FROM agent WHERE id > ? AND connection_status = ?;",
-    [WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE] = "SELECT id FROM agent WHERE id > ? AND connection_status = ? AND node_name = ?;",
+    [WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS] = "SELECT id FROM agent WHERE id > ? AND connection_status = ? ORDER BY id LIMIT ?;",
```
Comment (TomasTurina) [inline src/wazuh_db/wdb.c:177]: Undo changes to this query:
```suggestion
    [WDB_STMT_GLOBAL_GET_AGENTS_BY_CONNECTION_STATUS] = "SELECT id FROM agent WHERE id > ? AND connection_status = ?;",
```

File: src/wazuh_db/wdb_global.c [hunk 17]
```diff
@@ -1821,6 +1821,17 @@ cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id
             *status = WDBC_ERROR;
             return NULL;
         }
+        if (sqlite3_bind_int(stmt, 4, limit) != SQLITE_OK) {
+            merror("DB(%s) sqlite3_bind_int(): %s", wdb->id, sqlite3_errmsg(wdb->db));
+            *status = WDBC_ERROR;
+            return NULL;
+        }
+    } else {
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_global.c:1829]: Remove this else.

File: src/wazuh_db/wdb_parser.c [hunk 18]
```diff
@@ -5804,12 +5805,18 @@ int wdb_parse_global_get_agents_by_connection_status(wdb_t* wdb, char* input, ch
     }
     connection_status = next;
 
+    /* Get limit */
```
Comment (TomasTurina) [inline src/wazuh_db/wdb_parser.c:5808]: Get first the `node_name` and if it is not null, get the `limit`.
```c
    /* Get node name */
    next = strtok_r(NULL, delim, &savedptr);
    if (next != NULL) {
        node_name = next;

        /* Get limit */
        next = strtok_r(NULL, delim, &savedptr);
        if (next == NULL) {
            mdebug1("Invalid arguments 'limit' not found.");
            snprintf(output, OS_MAXSTR + 1, "err Invalid arguments 'limit' not found");
            return OS_INVALID;
        }
        limit = atoi(next);
    }
```

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/analysisd/asyscom.c [hunk 19]
```diff
@@ -23,22 +23,32 @@
...
     [ERROR_EMPTY_COMMAND] = "Empty command",
     [ERROR_UNRECOGNIZED_COMMAND] = "Unrecognized command",
     [ERROR_EMPTY_PARAMATERS] = "Empty parameters",
     [ERROR_EMPTY_SECTION] = "Empty section",
-    [ERROR_UNRECOGNIZED_SECTION] = "Unrecognized or not configured section"
+    [ERROR_UNRECOGNIZED_SECTION] = "Unrecognized or not configured section",
+    [ERROR_INVALID_AGENTS] = "Invalid agents parameter",
+    [ERROR_EMPTY_AGENTS] = "Error getting agents from DB",
+    [ERROR_EMPTY_LASTID] = "Empty last id",
+    [ERROR_TOO_MANY_AGENTS] = "Too many agents",
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/analysisd/asyscom.c:51]: Done!

File: src/wazuh_db/helpers/wdb_global_helpers.c [hunk 20]
```diff
@@ -37,7 +37,7 @@ static const char *global_db_commands[] = {
     [WDB_SET_AGENT_GROUPS] = "global set-agent-groups %s",
     [WDB_RESET_AGENTS_CONNECTION] = "global reset-agents-connection %s",
     [WDB_GET_AGENTS_BY_CONNECTION_STATUS] = "global get-agents-by-connection-status %d %s",
-    [WDB_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE] = "global get-agents-by-connection-status %d %s %s",
+    [WDB_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE] = "global get-agents-by-connection-status %d %s %d %s",
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.c:40]: Done!

File: src/wazuh_db/helpers/wdb_global_helpers.c [hunk 21]
```diff
@@ -1144,3 +1140,36 @@ time_t get_agent_date_added(int agent_id) {
...
+    int *array = NULL;
+    int len = 0;
+    wdbc_result status = WDBC_DUE;
+    char *node_name = NULL;
+    int aux_sock = -1;
+
+    node_name = get_node_name();
+    while (status == WDBC_DUE) {
+        // Query WazuhDB
+        snprintf(wdbquery, sizeof(wdbquery), global_db_commands[WDB_GET_AGENTS_BY_CONNECTION_STATUS_AND_NODE], last_id, connection_status, limit, node_name);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/helpers/wdb_global_helpers.c:1156]: Done!

File: src/wazuh_db/wdb.h [hunk 22]
```diff
@@ -2094,10 +2094,11 @@ int wdb_global_reset_agents_connection(wdb_t *wdb, const char *sync_status);
  * @param [in] connection_status Connection status of the agents requested.
  * @param [out] status wdbc_result to represent if all agents has being obtained or any error occurred.
  * @param [in] node_name Cluster node name
+ * @param [in] limit Limits the number of rows returned by the query.
  * @retval JSON with agents IDs on success.
  * @retval NULL on error.
  */
-cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status, const char* node_name);
+cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id, const char* connection_status, wdbc_result* status, const char* node_name, int limit);
```
[the remark this answers is not in the document]
Comment (sdvendramini) [reply src/wazuh_db/wdb.h:2116]: Done!

File: src/wazuh_db/wdb.c (hunk 16 above)
Comment (sdvendramini) [reply src/wazuh_db/wdb.c:180]: Done!

File: src/wazuh_db/wdb_parser.c (hunk 18 above)
Comment (sdvendramini) [reply src/wazuh_db/wdb_parser.c:5806]: Done!

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**